sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/rel/BeamCalcRel.java

```diff
@@ -68,15 +121,90 @@ public Calc copy(RelTraitSet traitSet, RelNode input, RexProgram program) {
+      final JavaTypeFactory typeFactory =
+          new JavaTypeFactoryImpl(BeamRelDataTypeSystem.INSTANCE) {
+            @Override
+            public Type getJavaClass(RelDataType type) {
+              if (type instanceof BasicSqlType || type instanceof IntervalSqlType) {
+                if (type.getSqlTypeName() == SqlTypeName.FLOAT) {
+                  return type.isNullable() ? Float.class : float.class;
+                }
+              }
+              return super.getJavaClass(type);
+            }
+          };
+      final BlockBuilder builder = new BlockBuilder();
+
+      final PhysType physType =
+          PhysTypeImpl.of(typeFactory, getRowType(), JavaRowFormat.ARRAY, false);
+
+      Expression input =
+          Expressions.convert_(Expressions.call(processContextParam, "element"), Row.class);
+
+      final RexBuilder rexBuilder = getCluster().getRexBuilder();
+      final RelMetadataQuery mq = RelMetadataQuery.instance();
+      final RelOptPredicateList predicates = mq.getPulledUpPredicates(getInput());
+      final RexSimplify simplify = new RexSimplify(rexBuilder, predicates, false, RexUtil.EXECUTOR);
+      final RexProgram program = BeamCalcRel.this.program.normalize(rexBuilder, simplify);
 
-      BeamSqlExpressionExecutor executor = new BeamSqlFnExecutor(BeamCalcRel.this.getProgram());
+      Expression condition =
+          RexToLixTranslator.translateCondition(
+              program,
+              typeFactory,
+              builder,
+              new InputGetterImpl(input, upstream.getSchema()),
+              null,
+              conformance);
 
-      Schema schema = CalciteUtils.toSchema(rowType);
-      PCollection<Row> projectStream =
-          upstream
-              .apply(ParDo.of(new CalcFn(executor, CalciteUtils.toSchema(rowType))))
-              .setRowSchema(schema);
-      projectStream.setRowSchema(CalciteUtils.toSchema(getRowType()));
+      List<Expression> expressions =
+          RexToLixTranslator.translateProjects(
+              program,
+              typeFactory,
+              conformance,
+              builder,
+              physType,
+              DataContext.ROOT,
+              new InputGetterImpl(input, upstream.getSchema()),
+              null);
+
+      // output = Row.withSchema(outputSchema)
+      Expression output = Expressions.call(Row.class, "withSchema", outputSchemaParam);
+      Method addValue = Types.lookupMethod(Row.Builder.class, "addValue", Object.class);
+
+      for (int index = 0; index < expressions.size(); index++) {
+        Expression value = expressions.get(index);
+        FieldType toType = outputSchema.getField(index).getType();
+
+        // .addValue(value)
+        output = Expressions.call(output, addValue, castOutput(value, toType));
+      }
+
+      // .build();
+      output = Expressions.call(output, "build");
+
+      // if (condition) {
+      //   c.output(output);
+      // }
+      builder.add(
```

Comment:
Some explanation of this and an example of the generated code would help here

Reply:
An example of the generated code is in the comment immediately above this block.